waveorder/io/reader.py

```diff
@@ -0,0 +1,158 @@
+import numpy as np
+import os
+import zarr
+from tifffile import TiffFile
+
+
+class MicromanagerReader:
+
+    def __init__(self, folder):
+        """
+        loads ome-tiff files in folder into zarr or numpy arrays
+        Strategy:
+            1. read all files in folder
+            2. search for the file whose omexml does not contain the element tag "BinaryOnly"-> this is the master ome-tiff
+            3. each series represents a micro-manager stage position.  upon call for data, assign this data to a dict
+
+        :param folder: folder containing all ome-tiff files
+        :param reader: tifffile or aicsimageio
+        """
+
+        self._positions = {}
+        self.mm_meta = None
+        self.stage_positions = 0
+        self.height = 0
+        self.width = 0
+        self.frames = 0
+        self.slices = 0
+        self.channels = 0
+        self.chnames = []
+
+        self.master_ome_tiff = self._get_master_ome_tiff(folder)
+        print(self.master_ome_tiff)
+        self._set_mm_meta()
+
+    def _set_mm_meta(self):
+        with TiffFile(self.master_ome_tiff) as tif:
+            self.mm_meta = tif.micromanager_metadata
+
+            if self.mm_meta['Summary']['Positions'] > 1:
+                self.stage_positions = []
+                for p in range(self.mm_meta['Summary']['Positions']):
+                    print(f"appending {self.mm_meta['Summary']['StagePositions'][p]}")
+                    self.stage_positions.append(self.mm_meta['Summary']['StagePositions'][p])
```

Comment:
We may want to do this in a smarter way.  Currently this will append the following dictionary for every position that would look like this: `{'Label': 'C1_p1', 'DefaultXYStage': 'XYStage', 'DefaultZStage': 'ZStage', 'GridRow': 0, 'GridCol': 0, 'DevicePositions': [{'Device': 'FocusDrive', 'Position_um': [6895.653104999999]}, {'Device': 'XYStage', 'Position_um': [27420.3, 6347.5]}, {'Device': 'ZStage', 'Position_um': [22.700000000000003]}]}`

What we can do here is a bit of filtering to clean up this dictionary and make it a bit easier to view.

Some Suggestions could be to make it so that the dictionary elements looks as such: 
`{'Label': 'C1_p1', 'GridRow': 0, 'GridCol': 0, 'DevicePositions': {'FocusDrive': [6895.653104999999]}, 'XYStage': [27420.3, 6347.5], 'Device': 'ZStage', 'Position_um': [22.700000000000003]}}`

Reply:
From what I can see, you're asking to reformat the `DevicePositions` from 
- a list of dictionaries, where each list element is a device dictionary, and each device dictionary contains `Device` and `Position_um` keys
to
- a dictionary, where each key is the device name (value of `Device` from above), and each value is the list of int/float (the `Position_um` from above)

In this case I might correct your suggestion to look like:
`{'Label': 'C1_p1', 'GridRow': 0, 'GridCol': 0, 'DevicePositions': {'FocusDrive': [6895.653104999999], 'XYStage': [27420.3, 6347.5], 'ZStage': [22.700000000000003] } }`

Then the access pattern changes from:
```
>>> self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][0]['Device']
'FocusDrive' 
>>> self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][0]['Position_um']
[6895.653104999999]
```
to
```
>>> self.mm_meta['Summary']['StagePositions'][p]['DevicePositions']['FocusDrive']
[6895.653104999999]
>>> self.mm_meta['Summary']['StagePositions'][p]['DevicePositions']['XYStage']
[27420.3, 6347.5]
```

One complication i see is that you have to know the device name ( `FocusDrive` or `XYStage` ) which can vary depending on the mm-config.  It's harder to programmatically search for a key that may or may not exist (which would return an error if does not exist).

It's easy enough to fix this dictionary structure.  There are also other things we can do -- write a method to parse values out and assign them to class attributes?

Reply:
You wouldn't have to search for the matching key, since it is already in the metadata. For every position, you can first store the names of the stage devices, and then loop through those names to find their values

```
devices = []
for i in range(len(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][0]):
    devices.append(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][i]['Device'])

stage_positions = []
for dev in devices:
    stage_positions.append({dev: float(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][dev]})
```

Reply:
> ```
> devices = []
> for i in range(len(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][0]):
>     devices.append(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][i]['Device'])
> 
> stage_positions = []
> for dev in devices:
>     stage_positions.append({dev: float(self.mm_meta['Summary']['StagePositions'][p]['DevicePositions'][dev]})
> ```

I see.  So the ask is to have stage_positions list contain only the device info, not `Label`, `Grid_row`, or `Grid_col`:
Is that right?

Before:
```
>>> stage_positions[0]
{'Label': 'C1_p1', 
'DefaultXYStage': 'XYStage', 
'DefaultZStage': 'ZStage', 
'GridRow': 0, 
'GridCol': 0, 
'DevicePositions': [{'Device': 'FocusDrive', 'Position_um': [6895.653104999999]}, 
                    {'Device': 'XYStage', 'Position_um': [27420.3, 6347.5]}, 
                    {'Device': 'ZStage', 'Position_um': [22.700000000000003]}
                    ]
}
```
after:
```
>>> stage_positions[0]
 {'FocusDrive': [6895.653104999999]},
  'XYStage': [27420.3, 6347.5], 
  'ZStage': [22.700000000000003]}